src/cdn.rs

```diff
@@ -59,50 +77,173 @@ impl CdnBackend {
+                ref invalidation_requests,
+                ..
+            } => {
                 let mut invalidation_requests = invalidation_requests
                     .lock()
                     .expect("could not lock mutex on dummy CDN");
 
-                invalidation_requests.extend(
-                    path_patterns
-                        .iter()
-                        .map(|p| (distribution_id.to_owned(), (*p).to_owned())),
+                let invalidation = CdnInvalidation {
+                    distribution_id: distribution_id.to_owned(),
+                    invalidation_id: caller_reference.to_string(),
+                    path_patterns: path_patterns.iter().cloned().map(str::to_owned).collect(),
+                };
+
+                invalidation_requests.push(invalidation.clone());
+                Ok(invalidation)
+            }
+        }
+    }
+
+    #[cfg(test)]
+    fn clear_active_invalidations(&self) {
+        match self {
+            CdnBackend::Dummy {
+                invalidation_requests,
+                ..
+            } => {
+                invalidation_requests
+                    .lock()
+                    .expect("could not lock mutex on dummy CDN")
+                    .clear();
+            }
+            CdnBackend::CloudFront { .. } => unreachable!(),
+        }
+    }
+
+    fn active_invalidations(&self, distribution_id: &str) -> Result<Vec<CdnInvalidation>, Error> {
+        match self {
+            CdnBackend::Dummy {
+                invalidation_requests,
+                ..
+            } => {
+                let invalidation_requests = invalidation_requests
+                    .lock()
+                    .expect("could not lock mutex on dummy CDN");
+
+                Ok(invalidation_requests
+                    .iter()
+                    .filter(|i| i.distribution_id == distribution_id)
+                    .cloned()
+                    .collect())
+            }
+            CdnBackend::CloudFront {
+                runtime, client, ..
+            } => Ok(
+                runtime.block_on(CdnBackend::list_active_cloudfront_invalidations(
+                    client,
+                    distribution_id,
+                ))?,
+            ),
+        }
+    }
+
+    #[instrument]
+    async fn list_active_cloudfront_invalidations(
+        client: &Client,
+        distribution_id: &str,
+    ) -> Result<Vec<CdnInvalidation>, Error> {
+        let mut stream = client
+            .list_invalidations()
+            .distribution_id(distribution_id)
+            .into_paginator()
+            .items()
+            .send();
+
+        let mut ids: Vec<String> = Vec::new();
+        while let Some(invalidation) = stream.next().await {
+            let invalidation = invalidation?;
+
+            if let (Some(id), Some(status)) = (invalidation.id(), invalidation.status()) {
+                // the `ListInvalidation` AWS API doesn't support filtering, so we
```

Comment:
What happens when we've got a pile of completed validations stacked up? Will we have to make tons of requests to the API, getting lists of completed validations?

Reply:
Yes, that's what would happen. 

the AWS SDK handles pagination for us.